### PR TITLE
Update controlsocket to user userservice

### PR DESCRIPTION
**Pull request by SimonRichardson**
The controlsocket used the legacy state package to add and remove users. This swaps over the origin of the user service from mongo to dqlite. Permissions aren't migrated, as that still requires the mongo database. Once we transition away from legacy state for permissions this should be updated.

Also, I swapped out the fake/stub mocks in the test for real mocks.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

### Manual

Check the worker is starting and the socket is being created:
```shell
$ make microk8s-operator-update
$ juju bootstrap [cloud] c
$ juju switch controller
$ juju ssh controller/0
$ stat /var/lib/juju/control.socket
```

Manually curl the socket:
```bash
$ curl -X POST http://a/metrics-users \
  -d '{"username":"juju-metrics-test","password":"bar"}' \
  --unix-socket /var/lib/juju/control.socket
# {"message":"created user \"juju-metrics-test\""}
```
```bash
$ curl -X DELETE http://a/metrics-users/juju-metrics-test \
  --unix-socket /var/lib/juju/control.socket
# {"message":"deleted user \"juju-metrics-test\""}
```

## Links

**Jira card:** JUJU-5691

### CODE REVIEW

File: agent/agentbootstrap/bootstrap.go [hunk 1]
```diff
@@ -231,6 +231,13 @@ func (b *AgentBootstrap) Initialize(ctx stdcontext.Context) (_ *state.Controller
 	// and a function to insert it into the database.
 	adminUserUUID, addAdminUser := userbootstrap.AddUserWithPassword(b.adminUser.Name(), auth.NewPassword(info.Password))
 
+	// Add initial metrics user to the database.
+	metricsPassword, err := password.RandomPassword()
```
Comment (manadart) [inline agent/agentbootstrap/bootstrap.go:235]: I assume we don't use this as part of bootstrap, so I'd have this in the bootstrap worker.

File: internal/worker/controlsocket/worker_test.go [hunk 2]
```diff
@@ -6,27 +6,31 @@ package controlsocket
...
+	"github.com/juju/juju/testing"
 )
 
 type workerSuite struct {
-	state  *fakeState
-	logger Logger
+	jujutesting.IsolationSuite
+
+	logger            Logger
+	userService       *MockUserService
```
Comment (manadart) [inline internal/worker/controlsocket/worker_test.go:32]: Good. Much prefer this to the fake.

File: internal/worker/controlsocket/worker.go [hunk 3]
```diff
@@ -34,9 +36,53 @@ const (
...
+	// AddUser will add a new user to the database and return the UUID of the
+	// user if successful. If no password is set in the incoming argument,
+	// the user will be added with an activation key.
+	// The following error types are possible from this function:
+	//   - usererrors.UserNameNotValid: When the username supplied is not valid.
+	//   - usererrors.AlreadyExists: If a user with the supplied name already exists.
+	//   - usererrors.CreatorUUIDNotFound: If a creator has been supplied for the user
+	//     and the creator does not exist.
+	//   - auth.ErrPasswordNotValid: If the password supplied is not valid.
+	AddUser(ctx context.Context, arg service.AddUserArg) (user.UUID, []byte, error)
```
Comment (manadart) [inline internal/worker/controlsocket/worker.go:53]: Let's add an slice of permissions to `AddUserArg`, and if not empty, create them in the same transaction.

File: agent/agentbootstrap/bootstrap.go (hunk 1 above)
Comment (SimonRichardson) [reply agent/agentbootstrap/bootstrap.go:235]: Sure I can change that.

File: internal/worker/controlsocket/worker.go (hunk 3 above)
Comment (SimonRichardson) [reply internal/worker/controlsocket/worker.go:53]: That would make my life easier, but it does make my argument that permissions and uses should be in the same domain.

Comment (SimonRichardson) [reply internal/worker/controlsocket/worker.go:53]: @hmlanigan ☝️

Comment (SimonRichardson) [reply internal/worker/controlsocket/worker.go:53]: See: https://github.com/juju/juju/pull/17067

File: internal/worker/controlsocket/worker.go [hunk 4]
```diff
@@ -157,23 +215,30 @@ func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Reques
...
-func (w *Worker) addMetricsUser(username, password string) (int, error) {
+func (w *Worker) addMetricsUser(ctx context.Context, username string, password auth.Password) (int, error) {
 	err := validateMetricsUsername(username)
 	if err != nil {
 		return http.StatusBadRequest, err
 	}
 
-	if password == "" {
-		return http.StatusBadRequest, errors.NotValidf("empty password")
+	creatorUser, err := w.userService.GetUserByName(ctx, userCreator)
```
Comment (Aflynn50) [inline internal/worker/controlsocket/worker.go:224]: Can this not be done outside the handler and have the creatorUser.UUID stored in the worker? Its getting the same user every time based on a constant string and only using its UUID.

File: internal/worker/controlsocket/worker.go [hunk 5]
```diff
@@ -157,23 +215,30 @@ func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Reques
...
+	if err != nil {
+		return http.StatusInternalServerError, errors.Annotatef(err, "retrieving creator user %q: %v", userCreator, err)
 	}
 
-	user, err := w.config.State.AddUser(username, username, password, userCreator)
+	_, _, err = w.userService.AddUser(ctx, service.AddUserArg{
```
Comment (Aflynn50) [inline internal/worker/controlsocket/worker.go:229]: Would there be any benifit doing `w.catacomb.Context(ctx)` here (and in other service calls) allow the worker kill method to bring this down too?

**Review by Aflynn50 — APPROVED** [review]
Very clean!

File: internal/worker/controlsocket/worker.go (hunk 4 above)
Comment (SimonRichardson) [reply internal/worker/controlsocket/worker.go:224]: I think this is ok for now, but you're right. Maybe revisit this in the future.

File: internal/worker/controlsocket/worker.go (hunk 5 above)
Comment (SimonRichardson) [reply internal/worker/controlsocket/worker.go:229]: The server is tied to the catacomb already, so if the catacomb dies, the server dies, thus the request dies.

**Pull request merged**